### PR TITLE
Add vote commands

### LINKED ISSUE #559: Add vote function in neo-cli

### PR DESCRIPTION
**Pull request by ProDog**
Close #559 
- add vote commands
- update protocol.json

### CODE REVIEW

File: neo-cli/CLI/MainService.Vote.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        [ConsoleCommand("register candidate", Category = "Vote Commands")]
+        private void OnRegisterCandidateCommand(UInt160 senderAccount, ECPoint publicKey)
+        {
+            if (NoWallet())
+            {
+                Console.WriteLine("Need open wallet!");
+                return;
+            }
+
+            ScriptBuilder scriptBuilder = new ScriptBuilder();
```
Comment (erikzhang) [inline neo-cli/CLI/MainService.Vote.cs:33]:
```suggestion
            using ScriptBuilder scriptBuilder = new ScriptBuilder();
```

Comment (erikzhang) [reply neo-cli/CLI/MainService.Vote.cs:33]: Please add `using` before all `ScriptBuilder` declarations.

File: neo-cli/CLI/MainService.Vote.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            {
+                scriptBuilder.EmitAppCall(NativeContract.NEO.Hash, "registerCandidate", publicKey);
+                script = scriptBuilder.ToArray();
+            }
+
+            SendTransaction(script, senderAccount);
+        }
+
+        /// <summary>
+        /// Process "get validators"
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:44]:
```suggestion
        /// Process "get candidates"
```

File: neo-cli/CLI/MainService.Vote.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+using System.Linq;
+using Neo.IO;
+using Neo.Cryptography.ECC;
+
+namespace Neo.CLI
+{
+    partial class MainService
+    {
+        /// <summary>
+        /// Process "register validator" command
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:20]:
```suggestion
        /// Process "register candidate" command
```

File: neo-cli/CLI/MainService.Vote.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                script = scriptBuilder.ToArray();
+            }
+
+            SendTransaction(script);
+        }
+
+        /// <summary>
+        /// Process "unregister validator" command
+        /// </summary>
+        /// <param name="senderAccount"></param>
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:62]: If there are no description for the arguments, i prefer to remove the `param name`

File: neo-cli/CLI/MainService.Vote.cs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            {
+                scriptBuilder.EmitAppCall(NativeContract.NEO.Hash, "getCandidates");
+                script = scriptBuilder.ToArray();
+            }
+
+            SendTransaction(script);
+        }
+
+        /// <summary>
+        /// Process "unregister validator" command
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:60]:
```suggestion
        /// Process "unregister candidate" command
```

File: neo-cli/CLI/MainService.Vote.cs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            List<Cosigner> signCollection = new List<Cosigner>();
+
+            if (account != null)
+            {
+                using (SnapshotView snapshot = Blockchain.Singleton.GetSnapshot())
+                {
+                    UInt160[] accounts = CurrentWallet.GetAccounts().Where(p => !p.Lock && !p.WatchOnly).Select(p => p.ScriptHash).Where(p => NativeContract.GAS.BalanceOf(snapshot, p).Sign > 0).ToArray();
+                    foreach (var signAccount in accounts)
+                    {
+
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:166]: Remove this line

File: neo-cli/CLI/MainService.Vote.cs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            using (ScriptBuilder scriptBuilder = new ScriptBuilder())
+            {
+                scriptBuilder.EmitAppCall(NativeContract.NEO.Hash, "getNextBlockValidators");
+                script = scriptBuilder.ToArray();
+            }
+
+            SendTransaction(script);
+        }
+
+        private void SendTransaction(byte[] script, UInt160 account = null)
```
Comment (shargon) [inline neo-cli/CLI/MainService.Vote.cs:155]: Maybe we should move this method to `MainService.cs`

Comment (ProDog) [reply neo-cli/CLI/MainService.Vote.cs]: Done

File: neo-cli/CLI/MainService.cs [hunk 8]
```diff
@@ -476,5 +479,59 @@ private static void WriteLineWithoutFlicker(string message = "", int maxWidth =
...
+            if (account != null)
+            {
+                using (SnapshotView snapshot = Blockchain.Singleton.GetSnapshot())
+                {
+                    UInt160[] accounts = CurrentWallet.GetAccounts().Where(p => !p.Lock && !p.WatchOnly).Select(p => p.ScriptHash).Where(p => NativeContract.GAS.BalanceOf(snapshot, p).Sign > 0).ToArray();
+                    foreach (var signAccount in accounts)
+                    {
+                        if (account.Equals(signAccount))
+                        {
+                            signCollection.Add(new Cosigner() { Account = signAccount });
```
Comment (vncoelho) [inline neo-cli/CLI/MainService.cs:496]: very good.

File: neo-cli/CLI/MainService.cs [hunk 9]
```diff
@@ -476,5 +479,59 @@ private static void WriteLineWithoutFlicker(string message = "", int maxWidth =
...
+                Transaction tx = CurrentWallet.MakeTransaction(script, account, null, signCollection?.ToArray());
+                Console.WriteLine($"Invoking script with: '{tx.Script.ToHexString()}'");
+
+                using (ApplicationEngine engine = ApplicationEngine.Run(tx.Script, tx, null, testMode: true))
+                {
+                    Console.WriteLine($"VM State: {engine.State}");
+                    Console.WriteLine($"Gas Consumed: {new BigDecimal(engine.GasConsumed, NativeContract.GAS.Decimals)}");
+                    Console.WriteLine($"Evaluation Stack: {new JArray(engine.ResultStack.Select(p => p.ToParameter().ToJson()))}");
+                    Console.WriteLine();
+                    if (engine.State.HasFlag(VMState.FAULT))
```
Comment (shargon) [inline neo-cli/CLI/MainService.cs:514]: I miss the deserialized result

**Comment by shargon** [comment]
Maybe you can check https://github.com/neo-project/neo-node/pull/575 and serialize the result according to the expected type?

**Comment by ProDog** [comment]
OK, I can complete it in a new PR.

**Pull request merged**